python/demo/medical_imaging/demo/demo_anisotropic_diffusion_medical.py

```diff
@@ -0,0 +1,293 @@
+    
+    sobel_orig = sobel_orig / (np.max(sobel_orig) + 1e-10)
+    sobel_denoised = sobel_denoised / (np.max(sobel_denoised) + 1e-10)
+    
+    numerator = np.sum(sobel_orig * sobel_denoised)
+    denominator = np.sqrt(np.sum(sobel_orig**2) * np.sum(sobel_denoised**2))
+    
+    epi = numerator / (denominator + 1e-10)
+    return epi
+
+
+def create_synthetic_test_image(output_file="data/synthetic_test.png", size=(512, 512)):
+    img = np.zeros(size, dtype=np.float64)
+    y, x = np.ogrid[:size[0], :size[1]]
+    
+    cx1, cy1, r1 = size[1]//4, size[0]//4, 60
+    mask1 = (x - cx1)**2 + (y - cy1)**2 <= r1**2
+    img[mask1] = 1.0
+    
+    cx2, cy2, r2 = 3*size[1]//4, 3*size[0]//4, 80
+    mask2 = (x - cx2)**2 + (y - cy2)**2 <= r2**2
+    img[mask2] = 0.9
+    
+    img[size[0]//3:2*size[0]//3, size[1]//2-40:size[1]//2+40] = 0.8
+    
+    np.random.seed(42)
+    noise = np.random.normal(0, 0.03, size)
+    noisy_img = np.clip(img + noise, 0, 1)
+    
+    img_uint8 = (noisy_img * 255).astype(np.uint8)
+    Path(output_file).parent.mkdir(parents=True, exist_ok=True)
+    Image.fromarray(img_uint8).save(output_file, quality=100, optimize=False)
+    
+    return output_file
+
+
+class AnisotropicDiffusion:
+    
+    def __init__(self, image_shape, kappa=20.0, dt=0.05, comm=MPI.COMM_WORLD):
+        self.image_shape = image_shape
+        self.kappa = kappa
+        self.dt = dt
+        self.comm = comm
+
+        self.domain = create_image_mesh(image_shape, comm)
+        self.V = fem.functionspace(self.domain, ("Lagrange", 1))
+        
+        self.u = ufl.TrialFunction(self.V)
+        self.v = ufl.TestFunction(self.V)
+        self.u_n = fem.Function(self.V)
+        
+    def diffusion_coefficient(self, grad_u):
+        grad_magnitude = ufl.sqrt(ufl.dot(grad_u, grad_u) + 1e-10)
+        c = 1.0 / (1.0 + (grad_magnitude / self.kappa)**2)
+        return c
+    
+    def setup_variational_problem(self):
+        c = self.diffusion_coefficient(ufl.grad(self.u_n))
+        
+        F = (self.u - self.u_n) * self.v * ufl.dx + \
+            self.dt * c * ufl.dot(ufl.grad(self.u), ufl.grad(self.v)) * ufl.dx
+        
+        a = ufl.lhs(F)
+        L = ufl.rhs(F)
+        
+        return a, L
+    
+    def solve_timestep(self, a, L):
+        A = assemble_matrix(fem.form(a))
+        A.assemble()
+        b = assemble_vector(fem.form(L))
+        b.ghostUpdate(addv=PETSc.InsertMode.ADD, mode=PETSc.ScatterMode.REVERSE)
+        
+        solver = PETSc.KSP().create(self.comm)
+        solver.setOperators(A)
+        solver.setType(PETSc.KSP.Type.CG)
+        
+        pc = solver.getPC()
+        pc.setType(PETSc.PC.Type.JACOBI)
+        
+        u_new = fem.Function(self.V)
+        solver.solve(b, u_new.x.petsc_vec)
+        u_new.x.scatter_forward()
+        
+        return u_new
+    
+    def denoise(self, image, n_iterations=30, verbose=True):
+        self.u_n = image_to_function(image, self.V)
```

Comment:
As your mesh aligns with the image grid, using interpolation to move data from a voxel grid (DG-0) to a P1 Lagrange space is not well defined. I would suggest that you for the first time step use `u_n` in DG-0, and then after solving, replace this function with a P1 function.